### PR TITLE
Added `getAll` operation

### PR DESCRIPTION
**Pull request by jabolina**
The `getAll` operation is used by the db-syncer to retrieve offset information.

### CODE REVIEW

File: manager/pom.xml [hunk 1]
```diff
@@ -47,6 +47,10 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-reactive-mysql-client</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-reactive-pg-client</artifactId>
+        </dependency>
```
Comment (jabolina) [inline manager/pom.xml]: We'll need this when running db-syncer with Postgres during tests. Adding in an additional commit.

Comment (ryanemerson) [reply manager/pom.xml]: We actually have a bit of a problem with multiple DB vendors atm. The DatabaseHandler injects a vertx `Pool` instance, however the implementation is tied to the `quarkus.datasource.db-kind` property which can only be configured at build time.

Eventually we will want different build profiles and/or modules per DB vendor, but this also complicates testing.

@wburns Any thoughts on how best to overcome this?

Comment (jabolina) [reply manager/pom.xml]: DOH, I thought db-kind could be overridden at runtime :disappointed:. To get things running I am OK removing this now and pinning the db-syncer to MySQL.

Comment (wburns) [reply manager/pom.xml]: This really needs to discussed outside of the context of this PR. We should just remove this change from here.

There are 2 options though I can think of. We have to have a different build per DB, or we can define multiple named providers and based on the url the user specifies at runtime we can pick out the appropriate one. The former will produce leaner and simpler code, the latter provides more flexibility.

File: manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java [hunk 2]
```diff
@@ -111,6 +113,26 @@ public void putAll(GingersnapHeader header, Map<byte[], byte[]> entryMap) {
       }
    }
 
+   public void getAll(GingersnapHeader header, Set<byte[]> keys) {
+      try {
+         String cacheName = header.getCacheName();
+         UniJoin.Builder<KeyValuePair<byte[], byte[]>> builder = Uni.join().builder();
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java]: After I thought about it, this needs to be done in a different way because https://github.com/gingersnap-project/cache-manager/pull/18/files#r1049741507 will also need it. This can be fine for a short term solution though, to get functionality working.

File: manager/pom.xml (hunk 1 above)
Comment (jabolina) [reply manager/pom.xml]: I removed the dependency for now.

File: manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java (hunk 2 above)
Comment (jabolina) [reply manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java]: I added the `putAll` to the `Caches` class.

**Comment by jabolina** [comment]
Rebasing.

**Comment by jabolina** [comment]
Rebased, just let me test again the putAll method

**Comment by jabolina** [comment]
Including a change that wraps the uses of `SearchBackend` in an `Instance`.

File: manager/src/main/java/io/gingersnapproject/Caches.java [hunk 3]
```diff
@@ -100,12 +122,22 @@ public Uni<String> put(String name, String key, String value) {
       return indexUni;
    }
 
+   public Uni<List<String>> putAll(String name, Map<String, String> values) {
+      // TODO use bulk operation to index
```
Comment (fax4ever) [inline manager/src/main/java/io/gingersnapproject/Caches.java]: Created https://github.com/gingersnap-project/cache-manager/issues/27

File: manager/pom.xml (hunk 1 above)
Comment (ryanemerson) [reply manager/pom.xml]: > This really needs to discussed outside of the context of this PR. We should just remove this change from here.

+1, I just wanted to make sure everyone was aware of the issue.

I have created https://github.com/gingersnap-project/cache-manager/issues/28 to continue the discussion and track the work.

**Review by fax4ever — COMMENTED** [review]
Nice job @jabolina! LGMT
I wanted to just add a test case covering the new 
1. getAll API 
2. putAll SPI

File: manager/src/main/java/io/gingersnapproject/search/SearchBackendFactory.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import javax.inject.Singleton;
+
+import io.gingersnapproject.search.opensearch.OpenSearchBackend;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.elasticsearch.client.RestClient;
+
+public class SearchBackendFactory {
+
+    @ConfigProperty(name = SearchBackend.PROPERTY, defaultValue = "null")
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/search/SearchBackendFactory.java]: Can this just not use the Instance field of the SearchBackend? I don't like doing our own factory logic here tbh. Also adding a new impl doesn't require any changes other than the new SearchBackend impl.

File: manager/src/main/java/io/gingersnapproject/Caches.java [hunk 5]
```diff
@@ -61,6 +71,16 @@ public Uni<String> put(String name, String key, String value) {
       return indexUni;
    }
 
+   public Uni<List<String>> putAll(String name, Map<String, String> values) {
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/Caches.java]: This should be in a separate PR imo. We need to make sure to have getAll in first.

File: manager/src/main/java/io/gingersnapproject/Caches.java [hunk 6]
```diff
@@ -80,6 +100,28 @@ public Stream<String> getKeys(String name) {
...
+      CacheAccessRecord<String> cacheAccessRecord = metrics.recordCacheAccess();
+      try {
+         Map<String, Uni<String>> res = getOrCreateMap(name).getAll(keys);
+
+         if (res.isEmpty()) return EMPTY_MAP_UNI;
+
+         UniJoin.Builder<KeyValuePair<String, String>> builder = Uni.join().builder();
+         for (Map.Entry<String, Uni<String>> entry : res.entrySet()) {
+            String key = entry.getKey();
+            cacheAccessRecord.localHit(entry.getValue() instanceof UniItem<String>);
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/Caches.java]: This won't work, we would need a separate cache access record per key.

Also this doesn't look like it completes the access. `uni.onItemOrFailure().invoke(cacheAccessRecord)`

File: manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java [hunk 7]
```diff
@@ -162,10 +164,51 @@ public void putAll(GingersnapHeader header, Map<byte[], byte[]> entryMap) {
...
             return;
          }
+         Map<String, String> transformed = new HashMap<>();
          for (Map.Entry<byte[], byte[]> entry : entryMap.entrySet()) {
-            maps.put(cacheName, toString(entry.getKey()), toString(entry.getValue()));
+            transformed.put(toString(entry.getKey()), toString(entry.getValue()));
          }
-         writeSuccess(header);
+
+         maps.putAll(cacheName, transformed).subscribe()
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java]: Lets do this in a separate PR. This has a bit of noise now.

File: manager/src/main/java/io/gingersnapproject/search/opensearch/OpenSearchBackend.java [hunk 8]
```diff
@@ -3,27 +3,25 @@
...
 public class OpenSearchBackend implements SearchBackend {
+   public static final String OPENSEARCH_ENABLED = "opensearch";
 
-   @Inject
-   RestClient restClient;
+   private final RestClient restClient;
 
    Set<String> mappedRules = ConcurrentHashMap.newKeySet();
 
+   public OpenSearchBackend(RestClient restClient) {
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/search/opensearch/OpenSearchBackend.java]: These changes are unneeded if we use Instance in the wrapper class instead.

File: manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java (hunk 7 above)
Comment (jabolina) [reply manager/src/main/java/io/gingersnapproject/hotrod/CommandProcessor.java]: Sure, I'll revert in here and open another.

File: manager/src/main/java/io/gingersnapproject/Caches.java (hunk 6 above)
Comment (jabolina) [reply manager/src/main/java/io/gingersnapproject/Caches.java]: I delegated this to the `get` operation.

**Comment by jabolina** [comment]
I reverted the `putAll` change, and I'll do it on another PR.

File: manager/src/test/java/io/gingersnapproject/hotrod/HotRodOperationsTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      if (cm != null) {
+         cm.stop();
+      }
+
+      cm = null;
+      cache = null;
+   }
+
+   @Test
+   public void testPutAll() {
```
Comment (wburns) [inline manager/src/test/java/io/gingersnapproject/hotrod/HotRodOperationsTest.java]:
```suggestion
   public void testGetAll() {
```

**Review by wburns — APPROVED** [review]
LGTM, just a simple change to test

**Comment by jabolina** [comment]
The `TimerMetric` needs to use the rule name during registration. Since every rule has the same name, the `MetricsResourceTest` is failing. Since my test does some operations, the metric of another rule is affected.

**Comment by wburns** [comment]
Integrated into main, the MetricsResourceTest is randomly failing for me as well. We need to fix that asap now though.
